mycroft/skills/skill_manager.py

```diff
@@ -12,543 +12,304 @@
-
-            # Update skills once an hour if update is enabled
-            if time.time() >= self.next_download and update:
-                self.download_skills()
-
-    def send_skill_list(self, message=None):
-        """
-            Send list of loaded skills.
-        """
+            self._reload_modified_skills()
+            self._load_new_skills()
+            self._unload_removed_skills()
+            self._update_skills()
+            sleep(2)  # Pause briefly before beginning next scan
+
+    def _remove_git_locks(self):
+        """If git gets killed from an abrupt shutdown it leaves lock files."""
+        for i in glob(os.path.join(self.msm.skills_dir, '*/.git/index.lock')):
+            LOG.warning('Found and removed git lock file: ' + i)
+            os.remove(i)
+
+    def _load_on_startup(self):
+        """Handle initial skill load."""
+        LOG.info('Loading installed skills...')
+        while not self.initial_load_complete:
+            skill_dirs = self._get_skill_directories()
+            if skill_dirs:
+                for skill_dir in skill_dirs:
+                    self._load_skill(skill_dir)
+                if len(self.skill_loaders) == len(skill_dirs):
+                    self.initial_load_complete = True
+            sleep(2)
+
+        LOG.info("Skills all loaded!")
+        self.bus.emit(Message('mycroft.skills.initialized'))
+
+    def _reload_modified_skills(self):
+        """Handle reload of recently changed skill(s)"""
+        for skill_dir in self._get_skill_directories():
+            skill_loader = self.skill_loaders.get(skill_dir)
+            if skill_loader is not None:
+                skill_loader.reload()
+
+    def _load_new_skills(self):
+        """Handle load of skills installed since startup."""
+        for skill_dir in self._get_skill_directories():
+            if skill_dir not in self.skill_loaders:
+                self._load_skill(skill_dir)
+
+    def _load_skill(self, skill_directory):
         try:
-            info = {}
-            for s in self.loaded_skills:
-                is_active = (self.loaded_skills[s].get('active', True) and
-                             self.loaded_skills[s].get('instance') is not None)
-                info[basename(s)] = {
-                    'active': is_active,
-                    'id': self.loaded_skills[s]['id']
-                }
-            self.bus.emit(Message('mycroft.skills.list', data=info))
-        except Exception as e:
-            LOG.exception(e)
-
-    def __deactivate_skill(self, skill):
-        """ Deactivate a skill. """
-        for s in self.loaded_skills:
-            if skill in s:
-                skill = s
-                break
+            skill_loader = SkillLoader(self.bus, skill_directory)
+            skill_loader.load()
+            self.skill_loaders[skill_directory] = skill_loader
+        except Exception:
+            LOG.exception('Load of skill {} failed!'.format(skill_directory))
```

Comment:
This causes an infinite loop in _load_on_startup if a skill fails to load. Maybe move the line
```python
            self.skill_loaders[skill_directory] = skill_loader
```

into a finally block

```python
    def _load_skill(self, skill_directory):
           skill_loader = SkillLoader(self.bus, skill_directory)
            skill_loader.load()
        except Exception:
            LOG.exception('Load of skill {} failed!'.format(skill_directory))
        finally:
            self.skill_loaders[skill_directory] = skill_loader
```